pkg/gameserverallocations/allocator.go

```diff
@@ -336,13 +342,19 @@ func (c *Allocator) allocateFromRemoteCluster(gsa *allocationv1.GameServerAlloca
 	request.MultiClusterSetting.Enabled = false
 	request.Namespace = connectionInfo.Namespace
 
+	ctx, cancel := context.WithTimeout(context.Background(), c.totalRemoteAllocationTimeout)
+	defer cancel() // nolint: errcheck
 	// Retry on remote call failures.
 	err = Retry(remoteAllocationRetry, func() error {
 		for i, ip := range connectionInfo.AllocationEndpoints {
+			select {
+			case <-ctx.Done():
+				return status.Errorf(codes.ResourceExhausted, "remote allocation retry timeout exceeded")
```

Comment:
This is the weird part. 
```
In Retry Loop I return ResourceExhausted error code as this is the only way to break out of Retry function. Not sure if you like this flow?
```
Logically it should be DeadlineExceeded but than `Retry` function will just keep retrying as it considers `DeadlineExceeded` a retryable error. The only error code that breaks you out of `Retry` is `ResourceExhausted` which makes sense I think.

Reply:
[Here is the logic](https://github.com/googleforgames/agones/blob/master/pkg/gameserverallocations/allocator.go#L558) that makes `ResourceExhausted` non-retryable.

However, `DeadlineExceeded` is a retryable error.  If the parent context's timeout has reached the **total** allocation timeout, even if the retry happens, it fails before attempting to remote allocate. The child timeout, should be retried for `DeadlineExceeded` error. We should add a unit test for the inner timeout and count the attempts.

Reply:
I am a bit confused here. Do you mean that I should return `DeadlineExceeded` error from select statement?
If that is the case I don't think it is the best approach because even though the remote allocate will not get executed we will still keep doing retries and will wait the full retry timeout as defined in `remoteAllocationRetry` variable.
I.e. my understanding is that if I change the code to:
`return status.Errorf(codes.ResourceExhausted, "remote allocation retry timeout exceeded")`
following will happen:
1) call out to `remoteAllocationCallback` happens and lets say it takes up the whole `totalRemoteAllocationTimeout`. At this point no future calls to allocate should be made and we should ideally exit `allocateFromRemoteCluster` function
2) `Retry` log get executed, so we wait for some period of time and do a retry
3) select statement is executed, we return `DeadlineExceeded`
4) `Retry` logic sees `DeadlineExceeded` waits exponential backoff and executes out logic again.
This will happen 7 times and take up a bunch of time as exponential backoff keeps getting longer.
Or have I misunderstood your comment?

Reply:
@kdima yes we should return `DeadlineExceeded`.
There are two contexts one is the parent context, with timeout `c.totalRemoteAllocationTimeout` and the second one is the child context created from the parent as `allocationCtx, cancel := context.WithTimeout(ctx, remoteAllocationTimeout)`. If the parent context is timed out, in the loop the select each time would go through the context.Done and not the default. If the child context is timed out, `grpcClient.Allocate` should return immediately. 

So in response to your comments:
No. "call out to remoteAllocationCallback" does not happen because each time in the loop ctx.Done will return the error. Since returning error is immediate we will not wait.

Please add a test case for inner timeout and count the attempts.

Reply:
You are correct the remoteAllocationCallBack will not be called, I missed that part. I still think that returning `DeadlineExceeded` results in the unexpected behaviour.
In this commit I have changed the error code to `DeadlineExceeded` and added debug print statement to my test case. In this test case the `totalRemoteAllocationTimeout` is set to 0, but the call to `allocateFromRemoteCluster` will always take approx 6 seconds. You can see that in the test case print out:
`Total time is 6.406179836s`
The reason it takes 6.4s seconds is that even though we return `DeadlineExceeded` the `Retry` logic will always retry 7 times, with timeouts starting from 100ms and exponential factor of 2 as defined in `remoteAllocationRetry` var. If you think that this is correct and expected behaviour I will keep the `DeadlineExceeded` error code. From my perspective this felt like unexpected behaviour.